compiler/rustc_hir_analysis/src/delegation.rs

```diff
@@ -275,26 +416,35 @@ pub(crate) fn inherit_predicates_for_delegation_item<'tcx>(
     tcx: TyCtxt<'tcx>,
     def_id: LocalDefId,
     sig_id: DefId,
-) -> Option<ty::GenericPredicates<'tcx>> {
+) -> ty::GenericPredicates<'tcx> {
     let args = create_generic_args(tcx, def_id, sig_id);
     let builder = PredicatesBuilder::new(tcx, args, sig_id);
+    let parent = tcx.parent(def_id.into());
 
     let caller_kind = fn_kind(tcx, def_id.into());
     let callee_kind = fn_kind(tcx, sig_id);
-
-    // FIXME(fn_delegation): Support generics on associated delegation items.
-    // Error will be reported in `check_constraints`.
     match (caller_kind, callee_kind) {
         (FnKind::Free, FnKind::Free)
         | (FnKind::Free, FnKind::AssocTrait) => {
-            Some(builder.build())
+            builder.build()
+        }
+
+        (FnKind::AssocTraitImpl, FnKind::AssocTrait) => {
+            builder
+            .with_parent(parent)
+            .with_inheritance_kind(InheritanceKind::Shallow)
```

Comment:
Maybe replace these methods with assignments?
```rust
builder.parent = parent;
builder.inh_kind = InheritanceKind::Shallow;
```
Not sure the builder pattern makes sense here.

Reply:
Builder pattern emphasizes the ctor semantics. Direct initialization looks unclear.

Reply:
Ok.